add index signature `[key:string]: any;` to Window

McFlat opened the pull request:
- https://stackoverflow.com/questions/42193262/element-implicitly-has-an-any-type-because-type-window-has-no-index-signatur#42193905
- https://basarat.gitbooks.io/typescript/docs/types/index-signatures.html

This is a pull request to fix a bug!

<!--
Thank you for submitting a pull request!

Here's a checklist you might find useful.
* [ ] There is an associated issue that is labeled
  'Bug' or 'help wanted' or is in the Community milestone
* [ ] Code is up-to-date with the `master` branch
* [ ] You've successfully run `jake runtests` locally
* [ ] You've signed the CLA
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

The reason for this pull request is to avoid having to set "suppressImplicitAnyIndexErrors":true in the tsconfig.json

Here's an example:
```
window[ 'VideoPlayer' ] = VideoPlayer
```

Fixes #27925

j-oliveras commented:
Files from `lib` folder are generated from `src/lib` folder. Some files from `src/lib` are generated from [TSJS-lib-generator](https://github.com/Microsoft/TSJS-lib-generator) repository.

Read [Contributing lib.d.ts fixes](https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md#contributing-libdts-fixes) for more info.

McFlat commented:
While looking at the files that are generated, it looks like this file doesn't seem to be generated at all.

j-oliveras commented:
`lib/lib.dom.d.ts` is generated from `src/lib/dom.generated.d.ts` and that is generated from TSJS-lib-generator repository.

McFlat commented:
Okay, in that case can you please link me to the exact file I have to modify then, which one is used to generate the `lib.dom.d.ts` file? So I can modify that file with the fix.

RyanCavanaugh commented:
This isn't something we'd accept in any form. Adding this index signature would make code like `window.getElmentById("foo")` typecheck without error, which would be really bad for overall usability.

McFlat commented:
@RyanCavanaugh we are trying to avoid writing buggy code and no use the `suppressImplicitAnyIndexErrors` setting in the tsconfig.json, this is the only thing in all of the code causing us to have to add that setting to tsconfig. I don't see how this change (which affects setting a index on the window) will make something like getting an element Id from the window typecheck without error. Please look closer at what we're working with here. Please also look at the stackoverflow link I attached. You want millions of developers to implement their own Window object just to avoid this bug? I don't see how your argument for why to close this pull request is valid.

Checkout this link, then please re-open and merge it.

https://stackoverflow.com/questions/42193262/element-implicitly-has-an-any-type-because-type-window-has-no-index-signatur#42193905

McFlat commented:
The issue/bug is a two part issue, I was able to resolve the first part and the change was merged into DefinitelyTyped, and the second part of the issue if this what I fixed in this pull request.

Here is the link to the merged code into DefinitelyTyped that I fixed with window, only thing now is to merge this code too. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29711

McFlat commented:
If the test passes, then I don't see the problem, it would fail in the use case you mention if that was true what you say, but it's not true, why don't you take another look at it. Too much pride?

McFlat commented:
> I don't think this attitude is going to help the conversation be productive.

Well at least you answer the question, instead of avoiding it. There's nothing wrong with asking a simple question, otherwise how else are people supposed to know. Avoiding a question says something too, so there's that, you didn't avoid it.

> Your fix doesn't even address the problem described in the StackOverflow issue because it modifies the wrong variable. If you had the "correct" change to allow the code described in the OP, then you would see the negative behavior I described.

How does it modify the wrong variable? It seems right to me, it adds the index settings capability to window object, just like you can in the browser console, you can set a `window['index'] = 1` and no error is thrown, all tests pass. Maybe if there was a erroneous behavior then the tests would fail. If that's the case then, How come your tests don't cover that erroneous use case you describe?

McFlat commented:
@RyanCavanaugh 

> Literally, this PR edits the wrong line of code in lib.d.ts. The behavior you want changed requires editing var Window instead of interface Window.

That's what I did I modified `var Window`, not `interface Window`.

<img width="343" alt="screen shot 2018-10-19 at 10 32 09 am" src="https://user-images.githubusercontent.com/43896676/47228247-5d5b1d00-d38a-11e8-8c01-81672dc64f7e.png">

RyanCavanaugh commented:
> Take a closer look please, and please enlighten me how you know what you're doing.

I made an error here and reversed the order of words here. You want to edit `interface`, not `var`.

McFlat commented:
Ok then say that, no problem I can make the change. Let's get this fixed so to not introduce any anti-patterns by making every project redefine the window interface.

RyanCavanaugh commented:
Again, that's not a change we would merge.

McFlat commented:
We can do `window.index = {}` instead, no problem, you can go take a nap now after all the frustration :)
